### PR TITLE
Add optional time window for GetTrace & SearchTrace of http_handler

### LINKED ISSUE #4150: [Feature]: time parameters for querying trace by id
### Requirement

add start_time and end_time as optional parameters in https://github.com/jaegertracing/jaeger-idl/blob/main/proto/api_v2/query.proto#L37

### Problem

when using with Granafa console, the trace id view takes a little long time to load.

image

This is because the query has to scan all traces in the db. we have about 4M sampled traces per day and the data is growing, and we store several days data in the database.

### Proposal

it would be good to add time parameters when querying by trace id. like tempo API https://grafana.com/docs/tempo/latest/api_docs/#query.

grafana console already have the time window. 

### Open questions

_No response_

### PR DESCRIPTION
**Pull request by rim99**
## Which problem is this PR solving?
Resolves #4150

## Description of the changes
- add optional time window when fetching trace by id

## How was this change tested?
- unittest

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: cmd/anonymizer/app/query/query.go [hunk 1]
```diff
@@ -49,15 +50,27 @@ func unwrapNotFoundErr(err error) error {
 }
 
 // QueryTrace queries for a trace and returns all spans inside it
-func (q *Query) QueryTrace(traceID string) ([]model.Span, error) {
+func (q *Query) QueryTrace(traceID string, startTime int64, endTime int64) ([]model.Span, error) {
```
Comment (yurishkuro) [inline cmd/anonymizer/app/query/query.go:53]: we always try to use strong types. Conversion from user-facing values can be done at the user interface layer, not at this internal API.
```suggestion
func (q *Query) QueryTrace(traceID string, startTime time.Time, endTime time.Time) ([]model.Span, error) {
```

File: storage/spanstore/interface.go [hunk 2]
```diff
@@ -50,6 +50,13 @@ type Reader interface {
 	FindTraceIDs(ctx context.Context, query *TraceQueryParameters) ([]model.TraceID, error)
 }
 
+// TraceGetParameters contains parameters of a trace get.
+type TraceGetParameters struct {
+	TraceID   model.TraceID
+	StartTime *time.Time
+	EndTime   *time.Time
```
Comment (yurishkuro) [inline storage/spanstore/interface.go:57]: time.Time has `IsZero` that can be used as an indication of "value not provided", instead of pointers.

File: storage/spanstore/interface.go [hunk 3]
```diff
@@ -50,6 +50,13 @@ type Reader interface {
 	FindTraceIDs(ctx context.Context, query *TraceQueryParameters) ([]model.TraceID, error)
 }
 
+// TraceGetParameters contains parameters of a trace get.
+type TraceGetParameters struct {
```
Comment (yurishkuro) [inline storage/spanstore/interface.go:54]: the name reads weird
```suggestion
type GetTraceParameters struct {
```

Comment (rim99) [reply storage/spanstore/interface.go:54]: I'm trying to follow the silimar naming pattern here:
- TraceQueryParameters
- OperationQueryParameters

These are like <Target>-<Verb>-Parameters

**Comment by yurishkuro** [comment]
please rebase

File: storage/spanstore/interface.go (hunk 3 above)
Comment (yurishkuro) [reply storage/spanstore/interface.go:54]: GetTraceParameters is the name used in the IDL

**Comment by rim99** [comment]
@yurishkuro Updated bsaed on comments, please reivew this PR, thanks

File: Makefile.Protobuf.mk [hunk 4]
```diff
@@ -14,9 +14,10 @@
 # instead of the go_package's declared by the imported protof files.
 #
 
-DOCKER_PROTOBUF_VERSION=0.5.0
-DOCKER_PROTOBUF=jaegertracing/protobuf:$(DOCKER_PROTOBUF_VERSION)
-PROTOC := docker run --rm -u ${shell id -u} -v${PWD}:${PWD} -w${PWD} ${DOCKER_PROTOBUF} --proto_path=${PWD}
+CONTAINER=docker
+CONTAINER_PROTOBUF_VERSION=0.5.0
+CONTAINER_PROTOBUF=jaegertracing/protobuf:$(CONTAINER_PROTOBUF_VERSION)
+PROTOC := ${CONTAINER} run --rm -u ${shell id -u} -v${PWD}:${PWD} -w${PWD} ${CONTAINER_PROTOBUF} --proto_path=${PWD}
```
Comment (yurishkuro) [inline Makefile.Protobuf.mk:20]:
```suggestion
DOCKER=docker
DOCKER_PROTOBUF_VERSION=0.5.0
DOCKER_PROTOBUF=jaegertracing/protobuf:$(DOCKER_PROTOBUF_VERSION)
PROTOC :=$(DOCKER) run --rm -u ${shell id -u} -v${PWD}:${PWD} -w${PWD} ${DOCKER_PROTOBUF} --proto_path=${PWD}
```

File: plugin/storage/grpc/proto/storage.proto [hunk 5]
```diff
@@ -72,6 +72,14 @@ message GetTraceRequest {
       (gogoproto.customtype) = "github.com/jaegertracing/jaeger/model.TraceID",
       (gogoproto.customname) = "TraceID"
     ];
+    google.protobuf.Timestamp start_time = 2 [
+      (gogoproto.stdtime) = true,
+      (gogoproto.nullable) = true
```
Comment (yurishkuro) [inline plugin/storage/grpc/proto/storage.proto:77]:
```suggestion
      (gogoproto.nullable) = false
```

File: plugin/storage/grpc/proto/storage.proto [hunk 6]
```diff
@@ -72,6 +72,14 @@ message GetTraceRequest {
...
+    ];
+    google.protobuf.Timestamp end_time = 3 [
+      (gogoproto.stdtime) = true,
+      (gogoproto.nullable) = true
```
Comment (yurishkuro) [inline plugin/storage/grpc/proto/storage.proto:81]:
```suggestion
      (gogoproto.nullable) = false
```

File: plugin/storage/grpc/proto/storage.proto [hunk 7]
```diff
@@ -72,6 +72,14 @@ message GetTraceRequest {
       (gogoproto.customtype) = "github.com/jaegertracing/jaeger/model.TraceID",
       (gogoproto.customname) = "TraceID"
     ];
+    google.protobuf.Timestamp start_time = 2 [
```
Comment (yurishkuro) [inline plugin/storage/grpc/proto/storage.proto:75]: please move this change to a separate PR

**Comment by yurishkuro** [comment]
@rim99 this PR is too large, please break it into logical pieces. At minimum grpc storage changes could be a separate PR, e.g. changing the IDL. The annonymizer change can be separate.

**Comment by yurishkuro** [comment]
you have jaeger-ui submodule change, it's not related

**Comment by rim99** [comment]
> @rim99 this PR is too large, please break it into logical pieces. At minimum grpc storage changes could be a separate PR, e.g. changing the IDL. The anonymizer change can be separate.

Cool, I'll keep creating new PRs and rebasing until we're satisfied with the size of this PR.

And frist PR is: https://github.com/jaegertracing/jaeger/pull/6242 . Other changes rely on this

File: Makefile.Protobuf.mk (hunk 4 above)
Comment (rim99) [reply Makefile.Protobuf.mk:20]: There is possibility that use this in cmd line

```
$(DOCKER}=podman
```

**Comment by rim99** [comment]
@yurishkuro Now this PR contains changes of http_handler only, please review

File: cmd/query/app/http_handler.go [hunk 8]
```diff
@@ -433,10 +439,21 @@ func (aH *APIHandler) getTrace(w http.ResponseWriter, r *http.Request) {
 	if !ok {
 		return
 	}
-	// TODO: add start time & end time
 	query := spanstore.GetTraceParameters{
 		TraceID: traceID,
 	}
+	startTime := r.FormValue(startTimeParam)
+	if startTime != "" {
+		// No need to handle err here, since time window is optional
```
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:447]: error handling is not about param being optional but about syntax errors. If we cannot parse the timestamp we need to let the user know why.

File: cmd/query/app/http_handler.go [hunk 9]
```diff
@@ -433,10 +439,21 @@ func (aH *APIHandler) getTrace(w http.ResponseWriter, r *http.Request) {
 	if !ok {
 		return
 	}
-	// TODO: add start time & end time
 	query := spanstore.GetTraceParameters{
 		TraceID: traceID,
 	}
+	startTime := r.FormValue(startTimeParam)
+	if startTime != "" {
```
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:446]: no need to leak variable to outer scope
```suggestion
	if startTime := r.FormValue(startTimeParam); startTime != "" {
```

File: cmd/query/app/http_handler.go [hunk 10]
```diff
@@ -466,11 +487,20 @@ func (aH *APIHandler) archiveTrace(w http.ResponseWriter, r *http.Request) {
...
+	}
+	endTime, err := aH.queryParser.parseTime(r, endTimeParam, time.Microsecond)
+	if aH.handleError(w, err, http.StatusBadRequest) {
+		return
+	}
 	query := spanstore.GetTraceParameters{
-		TraceID: traceID,
+		TraceID:   traceID,
+		StartTime: startTime,
+		EndTime:   endTime,
```
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:501]: from the start of the function up to this point it's the same code as in getTrace. Please move it to `parseGetTraceParameters` helper

File: cmd/query/app/http_handler_test.go [hunk 11]
```diff
@@ -388,6 +404,30 @@ func TestGetTraceBadTraceID(t *testing.T) {
...
+			name:  "Bad end time",
+			query: "end=b",
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ts := initializeTestServer(t)
+			var response structuredResponse
+			err := getJSON(ts.server.URL+`/api/traces/123456?`+tc.query, &response)
+			require.Error(t, err)
```
Comment (yurishkuro) [inline cmd/query/app/http_handler_test.go:426]: this is non-specific test, it can succeed for the other reasons than the condition you're testing. It's better to add `err string` to the test cases and check `require.ErrorContains`

File: cmd/query/app/http_handler_test.go [hunk 12]
```diff
@@ -411,6 +451,48 @@ func TestSearchByTraceIDSuccess(t *testing.T) {
...
+			query: "end=b",
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ts := initializeTestServer(t)
+			var response structuredResponse
+			err := getJSON(ts.server.URL+`/api/traces?traceID=1&`+tc.query, &response)
+			require.Error(t, err)
+		})
```
Comment (yurishkuro) [inline cmd/query/app/http_handler_test.go:492]: how is this test different from the one on L426?

File: cmd/query/app/http_handler.go (hunk 10 above)
Comment (rim99) [reply cmd/query/app/http_handler.go]: done

File: cmd/query/app/http_handler_test.go (hunk 11 above)
Comment (rim99) [reply cmd/query/app/http_handler_test.go:426]: done

File: cmd/query/app/http_handler_test.go (hunk 12 above)
Comment (rim99) [reply cmd/query/app/http_handler_test.go:504]: update with multiple traceIds

File: cmd/query/app/http_handler.go (hunk 8 above)
Comment (rim99) [reply cmd/query/app/http_handler.go]: done

File: cmd/query/app/http_handler_test.go (hunk 12 above)
Comment (rim99) [reply cmd/query/app/http_handler_test.go:501]: Actually, URIs are different:
- this is `/api/traces?traceID=1`, with a **?**
- the other is `/api/traces/traceID=1`, with a **/** in the same position

There are handled by different handler methods

File: cmd/query/app/http_handler.go [hunk 13]
```diff
@@ -425,18 +431,38 @@ func (aH *APIHandler) parseTraceID(w http.ResponseWriter, r *http.Request) (mode
...
+		}
+		return t, true
+	}
+	// It's OK if no time is specified, since it's optional
+	return time.Time{}, true
+}
+
+func (aH *APIHandler) parseGetTraceParameters(w http.ResponseWriter, r *http.Request) (spanstore.GetTraceParameters, bool) {
+	traceID, traceIdOk := aH.parseTraceID(w, r)
+	startTime, startTimeOk := aH.parseMicroseconds(w, r, startTimeParam)
```
Comment (yurishkuro) [inline cmd/query/app/http_handler.go:448]: you cannot defer checking the error like that because helper functions already called `aH.handleError` which writes to the client (you cannot do that more than once).

Comment (rim99) [reply cmd/query/app/http_handler.go]: I see, updated as checking result immediately

**Review by yurishkuro — APPROVED** [review]
